src/application/application.cpp

```diff
@@ -2,6 +2,15 @@
  * Copyright (c) 2016 Microsoft Corporation. All rights reserved.
  * Licensed under the MIT License. See LICENSE file in the project root for license information.
  */
+
+#include <string>
+#include <chrono>
+#include <cstdio>
+#include <ctime>
+#include <fstream>
+#include <sstream>
+#include <utility>
+
 #include <LightGBM/application.h>
```

Comment:
According to the Google style, `LightGBM/application.h` (the same issue with the most `cpp` files from this PR) should be the first one. Is there a bug in `cpplint` which blocks us from using the correct order? Refer to https://github.com/microsoft/LightGBM/pull/2066#issue-263709049.

Reply:
Oh thanks, I didn't know about #2066!

Yes, if I move just `LightGBM/application.h` back to the top and leave everything else the same in this PR, I get this from `cpplint`

> src/application/application.cpp:8:  Found C++ system header after other system header. Should be: application.h, c system, c++ system, other.  [build/include_order] [4]
src/application/application.cpp:9:  Found C++ system header after other system header. Should be: application.h, c system, c++ system, other.  [build/include_order] [4]
src/application/application.cpp:10:  Found C++ system header after other system header. Should be: application.h, c system, c++ system, other.  [build/include_order] [4]
src/application/application.cpp:11:  Found C++ system header after other system header. Should be: application.h, c system, c++ system, other.  [build/include_order] [4]
src/application/application.cpp:12:  Found C++ system header after other system header. Should be: application.h, c system, c++ system, other.  [build/include_order] [4]
src/application/application.cpp:13:  Found C++ system header after other system header. Should be: application.h, c system, c++ system, other.  [build/include_order] [4]
src/application/application.cpp:14:  Found C++ system header after other system header. Should be: application.h, c system, c++ system, other.  [build/include_order] [4]

I think it is probably because this is called `application/application.cpp` and `cpplint` is expecting to find `"application.h"` (header colocated with implementation). When I change `<LightGBM/application.h>` to `"application.h"` (and leave everything else in this PR the same), `cpplint` does not raise any warnings.

![image](https://user-images.githubusercontent.com/7608904/83365102-66d0a780-a36b-11ea-971e-38f29ebc7617.png)

If we want to stick to the intent of the style guide, I guess we could make (for example) `LightGBM/application.h` the first header and just add `//NOLINT`. That would be closer to what the style guide says, even if `cpplint` fails to pick it up. What do you think?

Reply:
> Yes, if I move just `LightGBM/application.h` back to the top and leave everything else the same in this PR, I get this from `cpplint`

Oh, it's a pity! But thanks, this is exactly what I wanted to know.

> What do you think?

I think that adding `//NOLINT` in each file is very ugly "fix" and makes no sense. Quoting myself from https://github.com/microsoft/LightGBM/pull/2066
> IMHO, any order is OK in case it's the same across the whole project.

I suggest only document our own new style guide for includes. So that we will be able to use it as a reference for new files in the future. Just like I did it earlier:

> But due to [the bug](https://github.com/cpplint/cpplint/issues/43) only the following order doesn't produce errors:
```
dir2/foo2.h.
A blank line
Your project's .h files in include folder.
A blank line
C system files.
C++ system files.
A blank line
Other libraries' .h files.
Your project's .h files in src folder.
```

Reply:
To be clear, do you mean I should push a commit adding this new standard somewhere? If so could you suggest where?

Reply:
Ah, no! I mean, just post it here as a comment.